security-analysis/security-analysis-api/src/main/java/com/powsybl/security/SecurityAnalysisResultBuilder.java

```diff
@@ -26,179 +26,198 @@
+        /**
+         * Adds a {@link LimitViolation} to the builder with a context.
+         * @param violation the context would be supplied to interceptors.
+         * @return
+         */
+        public B addViolation(LimitViolation violation, SecurityAnalysisResultContext limitViolationContext) {
+            Objects.requireNonNull(limitViolationContext);
+            violations.add(Objects.requireNonNull(violation));
+            interceptors.forEach(i -> i.onLimitViolation(violation, limitViolationContext));
+            return (B) this;
+        }
 
-        @Override
-        public void setComputationOk(boolean computationOk) {
-            this.computationOk = computationOk;
+        public B addViolations(List<LimitViolation> violations, SecurityAnalysisResultContext limitViolationContext) {
+            Objects.requireNonNull(violations).forEach(limitViolation -> addViolation(limitViolation, limitViolationContext));
+            return (B) this;
         }
 
-        @Override
-        public void addViolation(LimitViolation violation) {
-            violations.add(Objects.requireNonNull(violation));
+        public B addViolations(List<LimitViolation> violations) {
+            return addViolations(violations, resultContext);
         }
 
     }
 
-    private class PreContingencyResultBuilder extends LimitViolationsResultBuilder {
+    /**
+     * Builder for the pre-contingency result
+     */
+    public class PreContingencyResultBuilder extends AbstractLimitViolationsResultBuilder<PreContingencyResultBuilder> {
+
+        public PreContingencyResultBuilder(SecurityAnalysisResultContext resultContext) {
+            super(resultContext);
+        }
 
-        void endPreContingency() {
+        /**
+         * Finalize the creation of the PreContingencyResult instance
+         * @return the parent {@link SecurityAnalysisResultBuilder} instance.
+         */
+        public SecurityAnalysisResultBuilder endPreContingency() {
             List<LimitViolation> filteredViolations = filter.apply(violations, context.getNetwork());
             LimitViolationsResult res = new LimitViolationsResult(computationOk, filteredViolations);
-            interceptors.forEach(i -> i.onPreContingencyResult(context, res));
-            preContingencyResult = res;
+            interceptors.forEach(i -> i.onPreContingencyResult(res, resultContext));
+            setPreContingencyResult(res);
+
+            return SecurityAnalysisResultBuilder.this;
         }
     }
 
-    private class PostContingencyResultBuilder extends LimitViolationsResultBuilder {
+    public class PostContingencyResultBuilder extends AbstractLimitViolationsResultBuilder<PostContingencyResultBuilder> {
 
         private final Contingency contingency;
 
         PostContingencyResultBuilder(Contingency contingency) {
+            super(context);
+            this.contingency = Objects.requireNonNull(contingency);
+        }
```

Comment:
We should remove this method, and pass the context in the postContingency() method, to have a symmetrical design